### ISSUE TITLE
Errors resolving assets from multiple targets

### ISSUE THREAD

**Issue by ketulm** [sapphi-red/vite-plugin-static-copy#14]
The plugin crashes vite in dev mode when it doesn't find the requested asset in the first target.

```json
        "vite": "2.9.9",
        "vite-plugin-pwa": "0.12.0",
        "vite-plugin-static-copy": "0.5.0",
        "vite-raw-plugin": "1.0.1",
```

The requested asset exists in the second target.

```js
viteStaticCopy({
    targets: [
        { src: path.resolve(__dirname, 'node_modules/@myorg/test-library/src/assets/*'), dest: '.' },
        { src: 'src/assets/*', dest: '.' }
    ]
}),

// or 

viteStaticCopy({
    targets: [
        { src: [ path.resolve(__dirname, 'node_modules/@myorg/test-library/src/assets/*'), 
                 path.resolve(__dirname, 'src/assets/*') ],
          dest: '.'
        }
    ]
}),
```

In dev mode, vite process will crash when the asset from the second target is requested:

```
node:fs:1536
  handleErrorFromBinding(ctx);
  ^

Error: ENOENT: no such file or directory, stat '/Users/test/myproject/node_modules/@myorg/test-library/src/assets/test-icon.svg'
    at statSync (node:fs:1536:3)
    at sendStatic (/Users/test/myproject/node_modules/vite-plugin-static-copy/dist/index.cjs:665:45)
    at viteServeStaticCopyMiddleware (/Users/test/myproject/node_modules/vite-plugin-static-copy/dist/index.cjs:741:5)
    at call (/Users/test/myproject/node_modules/vite/dist/node/chunks/dep-59dc6e00.js:43823:7)
    at next (/Users/test/myproject/node_modules/vite/dist/node/chunks/dep-59dc6e00.js:43767:5)
    at Function.handle (/Users/test/myproject/node_modules/vite/dist/node/chunks/dep-59dc6e00.js:43770:3)
    at Server.app (/Users/test/myproject/node_modules/vite/dist/node/chunks/dep-59dc6e00.js:43635:37)
    at Server.emit (node:events:390:28)
    at parserOnIncoming (node:_http_server:951:12)
    at HTTPParser.parserOnHeadersComplete (node:_http_common:128:17) {
  errno: -2,
  syscall: 'stat',
  code: 'ENOENT',
  path: '/Users/test/myproject/node_modules/@myorg/test-library/src/assets/test-icon.svg'
}
```

Both targets have the same folder structure that we want to merge assets from when we build. The production build fails with this setup with the file already exists error. The current workaround is to use the `rollup-plugin-copy` but that lacks support for virtual static assets for dev-mode without copying them to project root.

```
[vite-plugin-static-copy:build] EEXIST: file already exists, mkdir '/Users/test/project/dist/images'
[vite-plugin-static-copy] Copy count was not set.

error during build:
Error: EEXIST: file already exists, mkdir '/Users/test/project/dist/images'
node:child_process:903
    throw err;
    ^

Error: Command failed: vite build --mode production
    at checkExecSyncError (node:child_process:826:11)
    at Object.execSync (node:child_process:900:15)
    at buildApp (/Users/test/project/build/build.js:7:11)
    at Object.<anonymous> (/Users/test/project/build/build.js:10:1)
    at Module._compile (node:internal/modules/cjs/loader:1095:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1147:10)
    at Module.load (node:internal/modules/cjs/loader:975:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:17:47 {
  status: 1,
  signal: null,
  output: [ null, null, null ],
  pid: 15469,
  stdout: null,
  stderr: null
}
```

For reference, the first `@myorg/test-library` is a private internal library that contains some common assets and components used by the parent `myproject`.

**Comment by toneb** [comment]
Problem seems to be in race condition in the copy loop, multiple copyTargets try to create the same folder.

Changing fs.copy to fs.copySync as a workaround fixes the above issue:
https://github.com/sapphi-red/vite-plugin-static-copy/blob/712a32740a9b584fd6ea1b574a9d89e7486244b6/src/utils.ts#L83

**Comment by stregouet** [comment]
I also get a race condition: when running `vite build` with a config similar to:

```js
viteStaticCopy({
    targets: [
        { src: path.resolve(__dirname, 'node_modules/@myorg/test-library/src/assets/*'), dest: 'assets' },
        { src: path.resolve(__dirname, 'node_modules/@myorg/other-library/src/assets/*'), dest: 'assets' },
    ]
}),
```
if both sources have a `favicon.ico` I got following errors:

```
[vite-plugin-static-copy:build] ENOENT: no such file or directory, unlink '/home/me/src/project/dist/assets/favicon.ico'
```
This comes from `utils.ts:copyAll`, `vite-plugin-static-copy` is trying to copy (with fs-extra copy function which will first unlink destination file and then copy it) all files in parallel. A workaround seems to do copy in sequential way:


```diff
diff --git a/src/utils.ts b/src/utils.ts
index 2104d0a..dd07595 100644
--- a/src/utils.ts
+++ b/src/utils.ts
@@ -72,19 +72,16 @@ export const copyAll = async (
   flatten: boolean
 ) => {
   const copyTargets = await collectCopyTargets(rootSrc, targets, flatten)
-  await Promise.all(
-    copyTargets.map(({ src, dest, transform }) => {
-      // use `path.resolve` because rootSrc/rootDest maybe absolute path
-      const resolvedSrc = path.resolve(rootSrc, src)
-      const resolvedDest = path.resolve(rootSrc, rootDest, dest)
-      if (transform) {
-        return transformCopy(transform, resolvedSrc, resolvedDest)
-      } else {
-        return fs.copy(resolvedSrc, resolvedDest)
-      }
-    })
-  )
-
+  for (const { src, dest, transform } of copyTargets) {
+    // use `path.resolve` because rootSrc/rootDest maybe absolute path
+    const resolvedSrc = path.resolve(rootSrc, src)
+    const resolvedDest = path.resolve(rootSrc, rootDest, dest)
+    if (transform) {
+      await transformCopy(transform, resolvedSrc, resolvedDest)
+    } else {
+      await fs.copy(resolvedSrc, resolvedDest)
+    }
+  }
   return copyTargets.length
 }
 ```

maybe with an option flag to keep parallel copy?

**Comment by sapphi-red** [comment]
@ketulm I have released a fix for this (v0.5.1). Does it work for you?

**Comment by ketulm** [comment]
@sapphi-red Yes, Thanks! It doesn't crash dev/build now. Missing assets also get handled correctly without crashing.
